flow/directives.py

```diff
@@ -0,0 +1,346 @@
+            return v
+
+        def default_finalize(v, directives):
+            return v
+
+        self._validator = identity if validator is None else validator
+        self._finalize = default_finalize if finalize is None else finalize
+
+    def __call__(self, value):
+        """Return a validated value for the given directive.
+
+        Parameters
+        ----------
+        value : any
+            The value to be validated by the :class:`~._DirectivesItem`. Accepted
+            values depend on the given instance. All `~._DirectivesItem`
+            instances support lazily validating callables of single (or multiple) jobs
+            that return an acceptable value for the given `~._DirectivesItem`
+            instance.
+
+        Returns
+        -------
+        any
+            Returns a immediately validated value for the given directive, or if
+            a callable was passed, a new callable is returned that wraps the
+            given callable with validator.
+        """
+        if callable(value):
+            @functools.wraps(value)
+            def validate_callable(*jobs):
+                return self._validator(value(*jobs))
+
+            return validate_callable
+        else:
+            return self._validator(value)
+
+
+class _Directives(MutableMapping):
+    """Class that maps environment and user specified objects for execution.
+
+    Resource request and execution style directives are the primary purpose for
+    this class. Examples include the number of processes expected to run or the
+    time to request to run.
+
+    Parameters
+    ----------
+    environment_directives : sequence of _DirectivesItem
+        The sequence of all environment-specified directives. All other
+        directives are user-specified and not validated. All environment
+        directives must be specified at initialization.
+    """
+
+    def __init__(self, environment_directives):
+        self._directive_definitions = dict()
+        self._defined_directives = dict()
+        self._user_directives = dict()
+
+        for directive in environment_directives:
+            self._add_directive(directive)
+
+    def _add_directive(self, directive):
+        if not isinstance(directive, _DirectivesItem):
+            raise TypeError(
+                f"Expected a _DirectivesItem object. Received {type(directive)}.")
+        elif directive._name in self._directive_definitions:
+            raise ValueError(f"Cannot re-define directive name {directive._name}.")
+        else:
+            self._directive_definitions[directive._name] = directive
+            self._defined_directives[directive._name] = directive._default
+
+    def _set_defined_directive(self, key, value):
+        try:
+            self._defined_directives[key] = self._directive_definitions[key](value)
+        except (KeyError, ValueError, TypeError) as err:
+            raise DirectivesError(f'Error setting directive "{key}"') from err
```

Comment:
We don't use quotes around the name/key in other error messages so I dropped the quotes here.
```suggestion
            raise DirectivesError(f'Error setting directive {key}') from err
```